pymeasure/instruments/anaheimautomation/dpseriesmotorcontroller.py

```diff
@@ -0,0 +1,377 @@
+
+    def step(self, steps, direction):
+        """Similar to the go() method, but also sets the number of steps and the direction in the same method.
+        
+        :param steps: Number of steps to clock
+        :param direction: Value to set on the direction property before sending the go command to the controller.
+                          Valid values of direction are those than can be set on the direction property ("CW" for
+                          clockwise or "CCW" for counter-clockwise).
+        """
+        self.steps = steps 
+        self.direction = direction
+        self.go()
+        
+    def slew(self, direction):
+        """Sends the slew command to the motor controller. This tells the controller to move the stepper motor until a
+           stop command is sent or a limit switch is reached.
+        
+        :param direction: value to set on the direction property before sending the slew command to the controller.
+        """
+        self.direction = direction
+        self.write("S") 
+    
+    def home(self, home_mode):
+        """ Send command to the motor controller to 'home' the motor.
+        
+        :param home_mode: 0 or 1 specifying which homing mode to run. 0 will perform a homing operation where the controller
+                          moves the motor until a soft limit is reached, then will ramp down to base speed and continue motion
+                          until a home limit is reached. In mode 1, the controller will move the motor until a limit is 
+                          reached, then will ramp down to base speed, change direction, and run until the limit is released.
+        """
+        hm = int(home_mode)
+        if hm == 0 or hm == 1:
+            self.write("H%i" % hm)
+        else:
+            raise ValueError("Invalid home mode %i specified!" % hm)
+    
+    def write(self, command):
+        """Override the instrument base write method to add the motor controller's id to the command string.
+        
+        :param command: command string to be sent to the motor controller.
+        """
+        # check if an address related command was sent. #
+        if "%" in command or "~" in command:
+            super().write("@%s" % command)
+        else:
+            super().write("@%i%s" % (self.address, command))
+
+    def values(self, command, **kwargs):
+        """ Override the instrument base values method to add the motor controller's id to the command string.
+        
+        :param command: command string to be sent to the motor controller.
+        """
+        # check if an address related command was sent. #
+        if "%" in command or "~" in command:
+            vals = super().values("@%s" % command, **kwargs)
+        else:
+            vals = super().values("@%i%s" % (self.address, command))
+
+        return vals
+    
+    def ask(self, command):
+        """ Override the instrument base ask method to add the motor controller's id to the command string.
```

Comment:
At first I was wondering why you'd redo this, then I realised that this surfaces an interesting inconsistency in our API - the VISAAdapter is the only one using a dedicated method for `ask` (PyVISA's `query`), the others just reuse their own `read()` and `write()` -- in those cases I fear you'll prepend the id twice. (You can maybe test with a `SerialAdapter`).
As a workaround, I think you should be able to check in your overridden `write` for `command.startswith("@")`, which means the command has already been preprocessed by `ask`.

Reply:
Good catch. Yeah when using the SerialAdapter the id was prepended twice. I added a check in write so that if @ is seen in the command string, then it just directly sends the command string as is.